File: validation/cases/pid_linear_step/pid_linear_step.m
% pid_linear_step.m -- Linear PI controller output via Octave Control Toolbox.
% Constructs matching PI transfer function and uses lsim() for open-loop comparison.
%
% ctrlpp uses backward Euler integration: I(z) = Ki*Ts*z/(z-1)
% Matching TF: C(z) = Kp + Ki*Ts*z/(z-1) = ((Kp + Ki*Ts)*z - Kp) / (z - 1)
%
% Usage: octave --no-gui pid_linear_step.m > pid_linear_step_octave.csv

pkg load control;

Kp = 2.0;
Ki = 1.0;
Kd = 0.0;
dt = 0.01;

% Build PI TF matching ctrlpp backward Euler convention
Ki_dt = Ki * dt;
C = tf([Kp + Ki_dt, -Kp], [1, -1], dt);

% Known error signal: decaying sinusoid
n_steps = 500;
t_vec = (0:n_steps-1)' * dt;
e = exp(-0.5 * t_vec) .* sin(2 * pi * 0.5 * t_vec);

% Compute controller output via lsim
[u_out] = lsim(C, e, t_vec);

fprintf('time,error,control\n');
for k = 1:n_steps
    fprintf('%.15e,%.15e,%.15e\n', t_vec(k), e(k), u_out(k));
end

File: validation/cases/pid_linear_step/pid_linear_step.cpp
// pid_linear_step.cpp -- Linear PI controller output matching pid_linear_step.m
// Open-loop: feeds known error signal to PID, compares control output.
// Usage: ./pid_linear_step > pid_linear_step_cpp.csv

#include "ctrlpp/control/pid.h"

#include <cmath>
#include <cstdio>
#include <numbers>

int main()
{
    using Pid = ctrlpp::pid<double, 1>;
    using Vec = Pid::vector_t;

    Pid::config_type cfg{};
    cfg.kp = Vec::Constant(2.0);
    cfg.ki = Vec::Constant(1.0);
    cfg.kd = Vec::Constant(0.0);

    Pid ctrl(cfg);

    constexpr double dt = 0.01;
    constexpr int n_steps = 500;

    std::printf("time,error,control\n");

    for(int k = 0; k < n_steps; ++k)
    {
        double t = k * dt;
        double e = std::exp(-0.5 * t) * std::sin(2.0 * std::numbers::pi * 0.5 * t);

        // PID compute: setpoint = e, measurement = 0  =>  error = e
        auto sp = Vec::Constant(e);
        auto meas = Vec::Constant(0.0);
        auto u = ctrl.compute(sp, meas, dt);

        std::printf("%.15e,%.15e,%.15e\n", t, e, u[0]);
    }
}
